packages/expo-dev-launcher/bundle/native-modules/DevMenuWebBrowser.ts

```diff
@@ -0,0 +1,91 @@
+import { NativeModules, AppState, Linking } from 'react-native';
+
+const DevMenu = NativeModules.ExpoDevMenuInternal;
+
+let redirectHandler: ((event: any) => void) | null = null;
+let onWebBrowserCloseAndroid: null | (() => void) = null;
+let isAppStateAvailable: boolean = AppState.currentState !== null;
+
+function onAppStateChangeAndroid(state: any) {
+  if (!isAppStateAvailable) {
+    isAppStateAvailable = true;
+    return;
+  }
+
+  if (state === 'active' && onWebBrowserCloseAndroid) {
+    onWebBrowserCloseAndroid();
+  }
+}
+
+function stopWaitingForRedirect() {
+  if (!redirectHandler) {
+    throw new Error(
+      `The WebBrowser auth session is in an invalid state with no redirect handler when one should be set`
+    );
+  }
+
+  Linking.removeEventListener('url', redirectHandler);
+  redirectHandler = null;
+}
+
+async function openBrowserAndWaitAndroidAsync(startUrl: string): Promise<any> {
+  const appStateChangedToActive = new Promise(resolve => {
+    onWebBrowserCloseAndroid = resolve;
+    AppState.addEventListener('change', onAppStateChangeAndroid);
+  });
+
+  let result = { type: 'cancel' };
+  await DevMenu.openWebBrowserAsync(startUrl);
+  const type = 'opened';
+
+  if (type === 'opened') {
```

Comment:
i'm a little confused about what's going on here 🤔 do you mean to be setting result.type instead?

Reply:
sorry these files are copied from the DevMenu package - I haven't really looked into them much yet

Reply:
I know that it looks very hacky and indeed it is. Btw. this code was mostly copied from `expo-web-browser`.